ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureClientUnitTest.kt

```diff
@@ -0,0 +1,819 @@
+        val cardinalClient = MockkCardinalClientBuilder().build()
+        val braintreeClient = MockkBraintreeClientBuilder().build()
+
+        val validateResponse = mockk<ValidateResponse> {
+            every { actionCode } returns CardinalActionCode.SUCCESS
+        }
+
+        every {
+            threeDSecureAPI.authenticateCardinalJWT(any(), any(), any())
+        } answers { call ->
+            val callback = call.invocation.args[2] as ThreeDSecureResultCallback
+            callback.onThreeDSecureResult(threeDSecureParams, null)
+        }
+
+        val sut = ThreeDSecureClient(
+            braintreeClient,
+            cardinalClient,
+            threeDSecureAPI,
+            merchantRepository
+        )
+
+        val paymentAuthResult = ThreeDSecurePaymentAuthResult("jwt", validateResponse, threeDSecureParams, null)
+        sut.tokenize(paymentAuthResult, threeDSecureTokenizeCallback)
+
+        val captor = slot<ThreeDSecureResult>()
+        verify { threeDSecureTokenizeCallback.onThreeDSecureResult(capture(captor)) }
+        val result = captor.captured
+        assertTrue(result is ThreeDSecureResult.Success)
+        assertEquals(result.nonce, threeDSecureParams.threeDSecureNonce)
+
+        verify { braintreeClient.sendAnalyticsEvent(ThreeDSecureAnalytics.JWT_AUTH_SUCCEEDED) }
+        verify { braintreeClient.sendAnalyticsEvent(ThreeDSecureAnalytics.VERIFY_SUCCEEDED) }
+    }
+
+    @Test
+    fun tokenize_whenValidateResponseSuccess_onAuthenticateCardinalJWTResultWithError_returnsResultAndSendsAnalytics() {
+        val cardinalClient = MockkCardinalClientBuilder().build()
+        val braintreeClient = MockkBraintreeClientBuilder().build()
+
+        val validateResponse = mockk<ValidateResponse> {
+            every { actionCode } returns CardinalActionCode.SUCCESS
+        }
+
+        val threeDSecureParams = mockk<ThreeDSecureParams>(relaxed = true) {
+            every { hasError() } returns true
+        }
+
+        every {
+            threeDSecureAPI.authenticateCardinalJWT(any(), any(), any())
+        } answers { call ->
+            val callback = call.invocation.args[2] as ThreeDSecureResultCallback
+            callback.onThreeDSecureResult(threeDSecureParams, null)
+        }
+
+        val sut = ThreeDSecureClient(
+            braintreeClient,
+            cardinalClient,
+            threeDSecureAPI,
+            merchantRepository
+        )
+
+        val paymentAuthResult = ThreeDSecurePaymentAuthResult("jwt", validateResponse, threeDSecureParams, null)
+        sut.tokenize(paymentAuthResult, threeDSecureTokenizeCallback)
+
+        val captorList = mutableListOf<ThreeDSecureResult>()
+        verify { threeDSecureTokenizeCallback.onThreeDSecureResult(capture(captorList)) }
+        val result = captorList.first()
+        assertTrue(result is ThreeDSecureResult.Failure)
+        assertEquals(result.nonce, paymentAuthResult.threeDSecureParams!!.threeDSecureNonce)
```

Comment:
```suggestion
        assertEquals(result.nonce, paymentAuthResult.threeDSecureParams?.threeDSecureNonce)
```

Instead of throwing a NPE, we can use the [safe call operator](https://kotlinlang.org/docs/null-safety.html#safe-call-operator) here. If `paymentAuthResult.threeDSecureParams` is null, null will be passed into the second parameter of `assertEquals` which would result in a clearer log if the assertion failed.

Reply:
Ah nice! Great to know!